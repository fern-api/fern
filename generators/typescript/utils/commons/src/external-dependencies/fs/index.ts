export { type Fs } from "./Fs";
