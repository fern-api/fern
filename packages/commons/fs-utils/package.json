{
  "name": "@fern-api/fs-utils",
  "version": "0.0.0",
  "repository": {
    "type": "git",
    "url": "https://github.com/fern-api/fern.git",
    "directory": "packages/commons/fs-utils"
  },
  "files": [
    "lib"
  ],
  "type": "module",
  "source": "src/index.ts",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "sideEffects": false,
  "scripts": {
    "clean": "rm -rf ./lib && tsc --build --clean",
    "compile": "tsc --build",
    "test": "vitest --passWithNoTests --run",
    "test:update": "vitest --passWithNoTests --run -u",
    "lint:eslint": "eslint --max-warnings 0 . --ignore-pattern=../../../.eslintignore",
    "lint:eslint:fix": "yarn lint:eslint --fix",
    "format": "prettier --write --ignore-unknown --ignore-path ../../../shared/.prettierignore \"**\"",
    "format:check": "prettier --check --ignore-unknown --ignore-path ../../../shared/.prettierignore \"**\"",
    "depcheck": "depcheck"
  },
  "dependencies": {
    "@fern-api/core-utils": "workspace:*",
    "@fern-api/path-utils": "workspace:*",
    "glob": "^11.0.0",
    "json-stream-stringify": "^3.1.4",
    "stream-json": "^1.8.0",
    "tmp-promise": "^3.0.3"
  },
  "devDependencies": {
    "@fern-api/core-utils": "workspace:*",
    "@fern-api/path-utils": "workspace:*",
    "@fern-api/configs": "workspace:*",
    "@types/node": "18.15.3",
    "depcheck": "^1.4.7",
    "eslint": "^8.56.0",
    "prettier": "^3.4.2",
    "@trivago/prettier-plugin-sort-imports": "^5.2.1",
    "typescript": "5.7.2",
    "@types/jest": "^29.5.14",
    "@types/stream-json": "^1.7.7",
    "vitest": "^2.1.8"
  }
}