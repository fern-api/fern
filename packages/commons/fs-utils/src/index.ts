export { AbsoluteFilePath } from "./AbsoluteFilePath";
export { cwd } from "./cwd";
export { dirname } from "./dirname";
export { doesPathExist, doesPathExistSync } from "./doesPathExist";
export {
    getDirectoryContents,
    type Directory,
    type File,
    type FileOrDirectory,
    getDirectoryContentsForSnapshot
} from "./getDirectoryContents";
export { join } from "./join";
export { listFiles } from "./listFiles";
export { moveFile } from "./moveFile";
export { moveFolder } from "./moveFolder";
export { relative } from "./relative";
export { RelativeFilePath } from "./RelativeFilePath";
export { relativize } from "./relativize";
export { resolve } from "./resolve";
export { streamObjectToFile } from "./streamObjectToFile";
export { stringifyLargeObject } from "./stringifyLargeObject";
export { waitUntilPathExists } from "./waitUntilPathExists";
export { streamObjectFromFile } from "./streamObjectFromFile";
export {
    convertToOsPath,
    convertToFernHostAbsoluteFilePath,
    convertToFernHostRelativeFilePath
} from "./osPathConverter";
export { getAllFilesInDirectory } from "./getAllFilesInDirectory";
export { getFilename } from "./getFilename";
export { isURL } from "./isUrl";
export { isCI } from "./isCI";
export { splitPath } from "./splitPath";
