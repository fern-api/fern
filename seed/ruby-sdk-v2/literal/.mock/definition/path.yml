imports:
  root: __package__.yml

service:
  auth: false
  base-path: ""
  endpoints:
    send:
      path: /path/{id}
      method: POST
      path-parameters:
        id: literal<"123">
      response: root.SendResponse
      examples: 
        - path-parameters:
            id: "123"
          response:  
            body: 
              message: "The weather is sunny"
              status: 200
              success: true
      