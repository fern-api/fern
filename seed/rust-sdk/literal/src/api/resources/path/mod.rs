pub mod path;
pub use path::*;
