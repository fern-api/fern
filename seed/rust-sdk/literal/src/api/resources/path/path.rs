use crate::api::*;
use crate::{ApiError, ClientConfig, HttpClient, RequestOptions};
use reqwest::Method;

pub struct PathClient {
    pub http_client: HttpClient,
}

impl PathClient {
    pub fn new(config: ClientConfig) -> Result<Self, ApiError> {
        Ok(Self {
            http_client: HttpClient::new(config.clone())?,
        })
    }

    pub async fn send(
        &self,
        id: &String,
        options: Option<RequestOptions>,
    ) -> Result<SendResponse, ApiError> {
        self.http_client
            .execute_request(Method::POST, &format!("path/{}", id), None, None, options)
            .await
    }
}
