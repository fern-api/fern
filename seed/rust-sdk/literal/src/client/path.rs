use crate::error::ApiError;
use crate::types::*;
use reqwest::Client;

pub struct PathClient {
    pub client: Client,
    pub base_url: String,
}

impl PathClient {
    pub fn new(base_url: String) -> Self {
        Self {
            client: Client::new(),
            base_url,
        }
    }

    pub async fn send(&self, id: &String) -> Result<serde_json::Value, ApiError> {
        todo!()
    }

}
