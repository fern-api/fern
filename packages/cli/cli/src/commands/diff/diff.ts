import { readFile } from "fs/promises";
import semver from "semver";

import { AbsoluteFilePath, cwd, doesPathExist } from "@fern-api/fs-utils";
import { resolve } from "@fern-api/fs-utils";
import { IntermediateRepresentation, serialization } from "@fern-api/ir-sdk";
import { IntermediateRepresentationChangeDetector } from "@fern-api/ir-utils";
import { FernCliError } from "@fern-api/task-context";

import { CliContext } from "../../cli-context/CliContext";

export interface Result {
    bump: "major" | "minor" | "patch";
    nextVersion?: string;
    errors: string[];
}

export async function diff({
    context,
    from,
    to,
    fromVersion
}: {
    context: CliContext;
    from: string;
    to: string;
    fromVersion: string | undefined;
}): Promise<Result> {
    const detector = new IntermediateRepresentationChangeDetector();
    const change = await detector.check({
        from: await readIr({ context, filepath: from, flagName: "from" }),
        to: await readIr({ context, filepath: to, flagName: "to" })
    });
    const errors = change.errors.map((error) => error.message);
    if (fromVersion == null) {
        return {
            bump: change.bump,
            errors
        };
    }
    const nextVersion = semver.inc(fromVersion, change.bump);
    if (!nextVersion) {
        context.failWithoutThrowing(`Invalid current version: ${fromVersion}`);
        throw new FernCliError();
    }
    return {
        bump: change.bump,
        nextVersion,
        errors
    };
}

async function readIr({
    context,
    filepath,
    flagName
}: {
    context: CliContext;
    filepath: string;
    flagName: string;
}): Promise<IntermediateRepresentation> {
    const absoluteFilepath = AbsoluteFilePath.of(resolve(cwd(), filepath));
    if (!(await doesPathExist(absoluteFilepath, "file"))) {
        context.failWithoutThrowing(`File not found: ${absoluteFilepath}`);
        throw new FernCliError();
    }
    const ir = await readFile(absoluteFilepath, "utf-8");
    const parsed = serialization.IntermediateRepresentation.parse(JSON.parse(ir));
    if (!parsed.ok) {
        context.failWithoutThrowing(`Invalid --${flagName}; expected a filepath containing a valid IR`);
        throw new FernCliError();
    }
    return parsed.value;
}
