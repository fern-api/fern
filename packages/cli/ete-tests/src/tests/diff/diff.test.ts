import { readdir } from "fs/promises";
import path from "path";

import { AbsoluteFilePath, RelativeFilePath, join } from "@fern-api/fs-utils";

import { diff } from "./diff";

const BREAKING_FIXTURES_DIR = join(
    AbsoluteFilePath.of(__dirname),
    RelativeFilePath.of("fixtures"),
    RelativeFilePath.of("breaking")
);

const NON_BREAKING_FIXTURES_DIR = join(
    AbsoluteFilePath.of(__dirname),
    RelativeFilePath.of("fixtures"),
    RelativeFilePath.of("non-breaking")
);

it(
    "breaking",
    async () => {
        const breakingChangeDirs = await readdir(BREAKING_FIXTURES_DIR, { withFileTypes: true });
        for (const dir of breakingChangeDirs) {
            if (!dir.isDirectory()) {
                throw new Error(
                    `${dir.name} is not a directory. All entries in ${BREAKING_FIXTURES_DIR} must be directories.`
                );
            }
            const result = await diff({
                fixturePath: AbsoluteFilePath.of(path.join(BREAKING_FIXTURES_DIR, dir.name)),
                fromVersion: "1.1.0"
            });
            expect(result.stdout).toMatchSnapshot();
            expect(result.exitCode).toBe(1);
        }
    },
    {
        timeout: 20_000
    }
);

it(
    "non-breaking",
    async () => {
        const nonBreakingChangeDirs = await readdir(NON_BREAKING_FIXTURES_DIR, { withFileTypes: true });
        for (const dir of nonBreakingChangeDirs) {
            if (!dir.isDirectory()) {
                throw new Error(
                    `${dir.name} is not a directory. All entries in ${NON_BREAKING_FIXTURES_DIR} must be directories.`
                );
            }
            const result = await diff({
                fixturePath: AbsoluteFilePath.of(path.join(NON_BREAKING_FIXTURES_DIR, dir.name)),
                fromVersion: "1.1.0"
            });
            expect(result.stdout).toMatchSnapshot();
            expect(result.exitCode).toBe(0);
        }
    },
    {
        timeout: 20_000
    }
);
