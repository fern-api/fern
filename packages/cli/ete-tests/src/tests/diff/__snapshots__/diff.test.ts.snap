// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`breaking 1`] = `"{"bump":"major","nextVersion":"2.0.0"}"`;

exports[`non-breaking 1`] = `"{"bump":"minor","nextVersion":"1.2.0"}"`;
